File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/VfsRootPlaylists.java
/**
 * @file
 * @brief
 * @author vitamin.caig@gmail.com
 */
package app.zxtune.fs;

import android.content.Context;
import android.net.Uri;

import androidx.annotation.Nullable;

import java.io.FileNotFoundException;
import java.io.InputStream;
import java.util.List;

import app.zxtune.R;
import app.zxtune.playlist.ProviderClient;

final class VfsRootPlaylists extends StubObject implements VfsRoot {

  static final String SCHEME = "playlists";

  private final Context context;
  private final ProviderClient client;

  VfsRootPlaylists(Context context) {
    this.context = context;
    this.client = ProviderClient.create(context);
  }

  @Override
  public Uri getUri() {
    return rootUriBuilder().build();
  }

  @Override
  public String getName() {
    return context.getString(R.string.vfs_playlists_root_name);
  }

  @Override
  public String getDescription() {
    return context.getString(R.string.vfs_playlists_root_description);
  }

  @Override
  @Nullable
  public VfsObject getParent() {
    return null;
  }

  @Override
  public void enumerate(Visitor visitor) {
    for (String name : client.getSavedPlaylists(null).keySet()) {
      visitor.onFile(new PlaylistFile(name));
    }
  }

  private static Uri.Builder rootUriBuilder() {
    return new Uri.Builder().scheme(SCHEME);
  }

  @Override
  @Nullable
  public VfsObject resolve(Uri uri) {
    if (SCHEME.equals(uri.getScheme())) {
      final List<String> path = uri.getPathSegments();
      if (path.isEmpty()) {
        return this;
      } else if (path.size() == 1) {
        final String name = path.get(0);
        return new PlaylistFile(name);
      }
    }
    return null;
  }

  private class PlaylistFile extends StubObject implements VfsFile {

    private final String name;

    PlaylistFile(String name) {
      this.name = name;
    }

    @Override
    public Uri getUri() {
      return rootUriBuilder().appendPath(name).build();
    }

    @Override
    public String getName() {
      return name;
    }

    @Override
    public VfsObject getParent() {
      return VfsRootPlaylists.this;
    }

    @Override
    public String getSize() {
      return getDescription();
    }

    @Override
    @Nullable
    public Object getExtension(String id) {
      if (VfsExtensions.INPUT_STREAM.equals(id)) {
        return openStream();
      } else {
        return super.getExtension(id);
      }
    }

    @Nullable
    private InputStream openStream() {
      try {
        final String path = client.getSavedPlaylists(name).get(name);
        return path != null ? context.getContentResolver().openInputStream(Uri.parse(path)) : null;
      } catch (FileNotFoundException e) {
      }
      return null;
    }
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/VfsRootPlaylists.kt
/**
 * @file
 * @brief
 * @author vitamin.caig@gmail.com
 */
package app.zxtune.fs

import android.content.Context
import android.net.Uri
import app.zxtune.R
import app.zxtune.playlist.ProviderClient.Companion.create
import kotlinx.coroutines.runBlocking
import java.io.FileNotFoundException

internal class VfsRootPlaylists(private val context: Context) : StubObject(), VfsRoot {
    private val client by lazy { create(context) }

    override val uri: Uri
        get() = rootUriBuilder().build()

    override val name
        get() = context.getString(R.string.vfs_playlists_root_name)

    override val description
        get() = context.getString(R.string.vfs_playlists_root_description)

    override val parent
        get() = null

    override fun enumerate(visitor: VfsDir.Visitor) = runBlocking {
        client.getSavedPlaylists()?.let { entries ->
            for ((name, path) in entries) {
                visitor.onFile(PlaylistFile(name, path))
            }
        } ?: Unit
    }

    override fun resolve(uri: Uri): VfsObject? {
        if (SCHEME == uri.scheme) {
            val path = uri.pathSegments
            if (path.isEmpty()) {
                return this
            } else if (path.size == 1) {
                val name = path[0]
                return PlaylistFile(name)
            }
        }
        return null
    }

    private inner class PlaylistFile(override val name: String, private var path: String? = null) :
        StubObject(), VfsFile {
        override val uri: Uri
            get() = rootUriBuilder().appendPath(name).build()

        override val parent
            get() = this@VfsRootPlaylists

        override val size: String
            get() = description

        override fun getExtension(id: String) = when (id) {
            VfsExtensions.INPUT_STREAM -> openStream()
            else -> super.getExtension(id)
        }

        private fun openStream() = try {
            if (path == null) {
                path = runBlocking { client.getSavedPlaylists(name)?.get(name) }
            }
            path?.let { context.contentResolver.openInputStream(Uri.parse(it)) }
        } catch (e: FileNotFoundException) {
            null
        }
    }

    companion object {
        const val SCHEME: String = "playlists"

        private fun rootUriBuilder() = Uri.Builder().scheme(SCHEME)
    }
}
